iceoryx_examples/user_header/README.md

```diff
@@ -0,0 +1,337 @@
+
+This are the definitions for the C++ API:
+<!-- [geoffrey] [iceoryx_examples/user_header/user_header_and_payload_types.hpp] [user-header] -->
+```cpp
+struct Header
+{
+    uint64_t publisherTimestamp{0};
+};
+```
+
+<!-- [geoffrey] [iceoryx_examples/user_header/user_header_and_payload_types.hpp] [user-payload] -->
+```cpp
+struct Data
+{
+    uint64_t fibonacci{0};
+};
+```
+
+This are the definitions for the C API:
+<!-- [geoffrey] [iceoryx_examples/user_header/user_header_and_payload_types.h] [user-header] -->
+```c
+typedef struct
+{
+    uint64_t publisherTimestamp;
+} Header;
+```
+
+<!-- [geoffrey] [iceoryx_examples/user_header/user_header_and_payload_types.h] [user-payload] -->
+```c
+typedef struct
+{
+    uint64_t fibonacci;
+} Data;
+```
+
+### Publisher Typed C++ API
+
+At first, there are includes for the user-header and user-payload types
+and the iceoryx includes for publisher, runtime and signal handling.
+<!-- [geoffrey] [iceoryx_examples/user_header/publisher_cxx_api.cpp] [iceoryx includes] -->
+```cpp
+#include "user_header_and_payload_types.hpp"
+
+#include "iceoryx_posh/popo/publisher.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+#include "iceoryx_utils/posix_wrapper/signal_handler.hpp"
+```
+
+The `main` function contains a loop which periodically sends a Fibonacci number.
+In order to break this loop, a signal handler is used which sets an atomic boolean.
+<!-- [geoffrey] [iceoryx_examples/user_header/publisher_cxx_api.cpp] [signal handling] -->
+```cpp
+std::atomic<bool> killswitch{false};
+
+static void sigHandler(int sig IOX_MAYBE_UNUSED)
+{
+    // caught SIGINT or SIGTERM, now exit gracefully
+    killswitch = true;
+}
+```
+
+The signal handler is registered in the `main` function.
+<!-- [geoffrey] [iceoryx_examples/user_header/publisher_cxx_api.cpp] [register sigHandler] -->
+```cpp
+auto signalIntGuard = iox::posix::registerSignalHandler(iox::posix::Signal::INT, sigHandler);
+auto signalTermGuard = iox::posix::registerSignalHandler(iox::posix::Signal::TERM, sigHandler);
+```
+
+Next, the iceoryx runtime is initialized. With this call, the application will be registered at `RouDi`,
+the routing and discovery daemon.
+<!-- [geoffrey] [iceoryx_examples/user_header/publisher_cxx_api.cpp] [initialize runtime] -->
+```cpp
+constexpr char APP_NAME[] = "iox-cpp-user-header-publisher";
+iox::runtime::PoshRuntime::initRuntime(APP_NAME);
+```
+
+Now, we create the publisher. Unlike the other examples, this uses the second template parameter to define the user-header.
+This is the only change compared to the other examples with the typed C++ API.
+<!-- [geoffrey] [iceoryx_examples/user_header/publisher_cxx_api.cpp] [create publisher] -->
+```cpp
+// for the 1.0 release, the Publisher alias for the PublisherImpl does not have the second parameter for the Header,
+// therefore the PublisherImpl must be used directly
+iox::popo::PublisherImpl<Data, Header> publisher({"Example", "User-Header", "Timestamp"});
```

Comment:
I messed this up. I'm not quite sure if I should fix this in this PR or create a separate one together with the release_1.0 branch. What do you think?
The changes are source and binary compatible since I only need to add a template parameter with a default value. So
```cpp
template <typename T>
using Publisher = PublisherImpl<T>;
```
becomes 
```cpp
template <typename T, typename H = mepoo::NoUserHeader>
using Publisher = PublisherImpl<T, H>;
```

Reply:
I'd vote for fixing it alongside the example here on this branch. For me this does not yet justify a v1.0.1 release. But once we do a v1.x we should definitely add it, same like #746

Reply:
agree, but I would already create a PR for the `release_1.0` branch. This does not mean that we immediately do a release, but at least we cannot forget to add fix it later

Reply:
I'll do the same for #746